working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+
+One possible perspective is that an extension type corresponds to an
+abstract data type: There is an underlying representation, but we wish to
+restrict the access to that representation to a set of operations that are
+independent of the operations available on the representation. In other
+words, the extension type ensures that we only work with the representation
+in specific ways, even though the representation itself has an interface
+that allows us to do many other (wrong) things.
+
+It would be straightforward to enforce an added discipline like this by
+writing a wrapper class with the allowed operations as members, and
+working on a wrapper object rather than accessing `o` and its methods
+directly:
+
+```dart
+// Attempt to emulate the extension type using a class.
+
+class TinyJson {
+  // `representation` is assumed to be a nested list of numbers.
+  final Object representation;
+
+  TinyJson(this.representation);
+
+  Iterable<num> get leaves sync* {
+    var self = representation;
+    if (self is num) {
+      yield self;
+    } else if (self is List<dynamic>) {
+      for (Object element in self) {
+        yield* TinyJson(element).leaves;
+      }
+    } else {
+      throw "Unexpected object encountered in TinyJson value";
+    }
+  }
+}
+
+void main() {
+  TinyJson tiny = TinyJson(<dynamic>[<dynamic>[1, 2], 3, <dynamic>[]]);
+  print(tiny.leaves);
+  tiny.add("Hello!"); // Error.
+}
+```
+
+This is similar to the extension type in that it enforces the use of
+specific operations (here we only have one: `leaves`) and in general makes
+it an error to use instance methods of the representation (e.g., `add`).
+
+Creation of wrapper objects takes time and space, and in the case where we
+wish to work on an entire data structure we'd need to wrap each object as
+we navigate the data structure. For instance, we need to create a wrapper
+`TinyJson(element)` in order to invoke `leaves` recursively.
+
+In contrast, the extension type mechanism is zero-cost, in the sense that
+it does _not_ use a wrapper object, it enforces the desired discipline
+statically.
+
+Like extension methods, extension types are static in nature: An extension
+type declaration may declare some type parameters (just like the current
+extension declarations). The type parameters will be bound to a type which
+is determined by the static type of the receiver. Similarly, like extension
+methods, members of an extension type are resolved statically, i.e., if
+`tiny.leaves` is an invocation of an extension type getter `leaves`, then
+the declaration named `leaves` whose body is executed is determined at
+compile-time. There is no support for late binding of an extension method,
+and hence there is no notion of overriding. In return for this lack of
+expressive power, we get improved performance.
+
+Here is another example. It illustrates the fact that an extension type `E`
+with on-type `T` introduces a type `E` which is a supertype of `T`. This
```

Comment:
I'm not sure I'm comfortable with describing this as a supertype relationship.  It implies an ability modify the supertype graph of a type after the fact, which this doesn't really do. Also, I'm not sure that instance checks and casts are going to behave as if this is a (proper) supertype relationship.  I think in so far as we allow this implicit injection, we probably should just treat it directly in assignability.

Reply:
Filed https://github.com/dart-lang/language/issues/1465 with some specific concerns.

Reply:
The discussion in #1465 will be used to clarify this topic area and make some design decisions which can be used to update this document. I think this means that the discussion has moved over there, so we can resolve this thread.